devito/mpi/distributed.py

```diff
@@ -421,6 +440,197 @@ def _rebuild(self, shape=None, dimensions=None, comm=None):
                            comm or self.comm)
 
 
+class SubDomainDistributor(DenseDistributor):
+
+    """
+    Decompose a subdomain over a set of MPI processes.
+
+    Parameters
+    ----------
+    subdomain : SubDomain
+        The subdomain to be decomposed
+
+    Notes
+    -----
+    This class is used internally by Devito for distributing
+    Functions defined on SubDomains. It stores reference to the
+    parent Distributor from which it is created.
+    """
+
+    def __init__(self, subdomain):
+        # Does not keep reference to the SubDomain since SubDomain will point to
+        # this SubDomainDistributor and Distributor does not point to the Grid
+
+        super().__init__(subdomain.shape, subdomain.dimensions)
+
+        self._dimension_map = frozendict({pd: sd for pd, sd
+                                          in zip(subdomain.grid.dimensions,
+                                                 subdomain.dimensions)})
+
+        self._parent = subdomain.grid.distributor
+
+        self._comm = self.parent.comm
+        self._topology = self.parent.topology
+
+        # Want to calculate and keep track of crosses here
+        self.__align_parent__()
+
+    def __align_parent__(self):
+        """
+        Routines for determining alignment of the SubDomain which this
+        SubDomainDistributor decomposes relative to the overarching
+        decomposition of the grid.
+        """
+        dist_interval = {dim: Interval(s.start, s.stop-1)
+                         for dim, s in self.parent.glb_slices.items()}
+
+        # Assumes no override of x_m and x_M supplied to operator
+        bounds_map = {**{dim.symbolic_min: 0
+                         for dim in self.parent.dimensions},
+                      **{dim.symbolic_max: sha-1
+                         for dim, sha in zip(self.parent.dimensions,
```

Comment:
I seem to remember that making `SubDistributor.dimensions` not match `SubDomain.dimensions` requires a bunch of extra complexity anywhere it's accessed, so I want to leave these properties intact. Have introduced `par_dimensions` and `par_shape` for the latter two however, as I agree it would be tidier with the attributes.